src/query/service/src/sql/planner/binder/copy.rs

```diff
@@ -500,3 +506,123 @@ impl<'a> Binder {
         Ok(())
     }
 }
+
+/// Named stage(start with `@`):
+///
+/// ```sql
+/// copy into mytable from @my_ext_stage
+///     file_format = (type = csv);
+/// ```
+///
+/// Returns user's stage info and relative path towards the stage's root.
+///
+/// If input location is empty we will convert it to `/` means the root of stage
+///
+/// - @mystage => (mystage, "/")
+///
+/// If input location is endswith `/`, it's a folder.
+///
+/// - @mystage/ => (mystage, "/")
+///
+/// Otherwise, it's a file
+///
+/// - @mystage/abc => (mystage, "abc")
+///
+/// For internal stage, we will also add prefix `/stage/<stage>/`
+///
+/// - @internal/abc => (internal, "/stage/internal/abc")
+pub async fn parse_stage_location(
+    ctx: &Arc<dyn TableContext>,
+    location: &str,
+) -> Result<(UserStageInfo, String)> {
+    let s: Vec<&str> = location.split('@').collect();
+    // @my_ext_stage/abc/
+    let names: Vec<&str> = s[1].splitn(2, '/').filter(|v| !v.is_empty()).collect();
+    let stage = UserApiProvider::instance()
+        .get_stage(&ctx.get_tenant(), names[0])
+        .await?;
+
+    let path = names.get(1).unwrap_or(&"").trim_start_matches('/');
+
+    let prefix = stage.get_prefix();
+    let relative_path = format!("{prefix}{path}");
+
+    debug!("parsed stage: {stage:?}, path: {relative_path}");
+    Ok((stage, relative_path))
+}
+
+/// parse_stage_location_v2 work similar to parse_stage_location.
+///
+/// Difference is input location has already been parsed by parser.
+pub async fn parse_stage_location_v2(
+    ctx: &Arc<dyn TableContext>,
+    name: &str,
+    path: &str,
+) -> Result<(UserStageInfo, String)> {
+    debug_assert!(path.starts_with('/'), "path should starts with '/'");
```

Comment:
Why use `debug_assert` here?

Reply:
I just copied it from the old place. I plan to refactor them with unit tests.

Reply:
Tracked at https://github.com/datafuselabs/databend/issues/7772